src/fromager/models.py

```diff
@@ -0,0 +1,244 @@
+
+    model_config = MODEL_CONFIG
+
+    env: EnvVars = Field(default_factory=dict)
+    """Additional env vars (overrides package env vars)"""
+
+    index_url: HttpUrl | None = None
+    """Alternative package index for pre-built wheel"""
+
+    pre_built: bool = False
+    """Use pre-built wheel from index server?"""
+
+
+_DictStrAny = dict[str, typing.Any]
+
+
+class PackageSettings(BaseModel):
+    """Package settings
+
+    yaml::
+        build_dir: python
+        changelog:
+            "1.0.1":
+                - fixed bug
+        env:
+            EGG: spam
+        download_source:
+            url: https://egg.test
+            destination_filename: new_filename
+        resolver_dist:
+            sdist_server_url: https://sdist.test/egg
+            include_sdists: true
+            include_wheels: false
+        variants:
+            cpu:
+                env:
+                    EGG: spamalot
+                index_url: https://wheel.test/simple
+            rocm:
+                pre_built: True
+    """
+
+    model_config = MODEL_CONFIG
+
+    name: Package
+    """Canonicalized package name"""
+
+    build_dir: BuildDirectory | None = None
+    """sub-directory with setup.py or pyproject.toml"""
+
+    changelog: Changelog = Field(default_factory=dict)
+    """Changelog entries"""
+
+    env: EnvVars = Field(default_factory=dict)
+    """Common env var for all variants"""
+
+    download_source: DownloadSource = Field(default_factory=DownloadSource)
+    """Alternative source download settings"""
+
+    resolver_dist: ResolverDist = Field(default_factory=ResolverDist)
+    """Resolve distribution version"""
+
+    variants: dict[Variant, PackageVariant] = Field(default_factory=dict)
+    """Variant configuration"""
+
+    @field_validator("download_source", "resolver_dist", "variants", mode="before")
+    @classmethod
+    def before_none_dict(
+        cls, v: _DictStrAny | None, info: ValidationInfo
+    ) -> _DictStrAny:
+        if v is None:
+            v = {}
+        return v
+
+    @classmethod
+    def from_string(cls, package: str | Package, raw_yaml: str) -> "PackageSettings":
+        """Load from raw yaml string"""
+        package = Package(canonicalize_name(package, validate=True))
+        parsed: typing.Any = yaml.safe_load(raw_yaml)
+        if parsed is None:
+            # empty file
+            parsed = {}
+        elif not isinstance(parsed, dict):
+            raise TypeError(f"{package}: invalid yaml, not a dict: {parsed}")
+        return cls(name=package, **parsed)
+
+    @classmethod
+    def from_file(cls, filename: pathlib.Path) -> "PackageSettings":
+        """Load from file
+
+        Raises :exc:`FileNotFound` when the file is not found.
+        The package name is taken from the stem of the file name.
+        """
+        filename = filename.absolute()
+        raw_yaml = filename.read_text(encoding="utf-8")
+        return cls.from_string(filename.stem, raw_yaml)
```

Comment:
If this fails, the error message won't include a reference to the filename.
```suggestion
        try:
          return cls.from_string(filename.stem, raw_yaml)
        except Exception as err:
          raise RuntimeError(f"failed to load settings from {filename}: {err}") from err
```

Reply:
Good idea!